### DISCUSSION TITLE
Prevent pytest from running imported functions as tests

### DISCUSSION THREAD

**Discussion by nielsbuwen** [pytest-dev/pytest#11366] [Q&A] [answered]
It seems that pytest is running functions that are imported inside a test file as if they were test functions.

My directory structure:

```
root/
    module.py  # contains def test_stuff()
    tests/
        __init__py
        test_it.py  # from module import test_stuff
```

So my question: is there a way to prevent pytest from "collecting" `test_stuff`? I tried these hooks:

- `pytest_ignore_collect(collection_path, path, config)`: none of the paths mention `module.py`
- `pytest_collect_file(file_path, path, parent)`: none of the paths mention `module.py`
- `pytest_pycollect_makeitem(collector, name, obj)`: this is actually called for `test_stuff` but i cannot prevent the creation here, can i?
- `pytest_collection_modifyitems(session, config, items)`: here i was able to remove the collected item, but this has other problems

The problems with `pytest_collection_modifyitems` are:

- As `test_stuff` is actually a generator function i get a warning like "yield tests were removed in pytest 4.0". I could ignore that, it is hard to get right, because i don't know how many functions pytest will collect in my real project.
- The collection counter is still increased and will be off by one. This is a problem for me, because i want to provide a test framework to my users and they need the correct amount of executed tests. E.g. pytest says: "collected 2 items" when one of those two was already removed in `pytest_collection_modifyitems`

Are there other ways to prevent imported (outside of `tests/`) functions from being collected?

**Comment by The-Compiler** [comment, marked as answer]
Not really - there is no good way for pytest to know where something is defined from (short of maybe checking `.__module__`, but that is its own can of worms).

You'll either need to rename it, do a `from mod import test_stuff as other_name`, or set `__test__ = False` on it (not sure if that works on functions though).

**Reply by nielsbuwen** [reply]
`__test__ = False` seems to work on functions. Thanks for the tip, this is good enough for me at the moment
